activerecord/lib/active_record/transactions.rb

```diff
@@ -206,6 +206,19 @@ def transaction(options = {}, &block)
         connection.transaction(options, &block)
       end
 
+      def start_sandbox
+        @sandbox_connections = connection_handler.connection_pools
+          .map(&:connection)
+          .each {|c| c.begin_transaction joinable: false }
+      end
+
+      def finish_sandbox
+        if instance_variable_defined?(:@sandbox_connections) && @sandbox_connections
+          @sandbox_connections.each {|c| c.rollback_transaction if c.transaction_open? }
+          @sandbox_connections = nil
+        end
+      end
+
```

Comment:
Can extract this implementation and state-tracking to a new `ActiveRecord::Sandbox` class.